### PR TITLE
option: reduce default number for TCP CT max entries

### PR DESCRIPTION
**Pull request by tklauser**
Commit e824a86bba21 ("daemon: Allow configuration of CT max entries")
bumped the default value to 1000000 in order to ease upgrades from
Cilium 1.2. In the helm charts, the value was again set to 512KB via the
`ct-global-max-entries-tcp` option. However, if Cilium is not deployed
via helm charts (e.g. when running as a systemd service in the devel
VM) the large default number of entries is used.

Set the default value for `bpf-ct-global-tcp-max` to 512KB again and
instead advise users in the helm chart comments to set it to 1000000 in
case they're upgrading or changed the size manually using Helm.

Since the default value of `bpf-nat-global-max` for the NAT table size
is derived from the default for `bpf-ct-global-tcp-max`, this commit
will also decrease the the NAT table size to 512K.
    
Document possible consequences of upgrading Cilium installations with
larger TCP CT and NAT table sizes.

This saves about ~150MB of memory at runtime.

Updates #10056

```release-note
The default maximum number of entries in the BPF TCP ctmap is reduced to 512K.
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/10289)

### CODE REVIEW

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-docs-please

File: Documentation/cmdref/cilium-agent.md
Comment (aanm) [inline Documentation/cmdref/cilium-agent.md]: I think there was a reason for this. @joestringer do you remember?

Comment (joestringer) [reply Documentation/cmdref/cilium-agent.md]: Any time we change this setting, we need to be very clear to users as this can cause connection disruption during upgrade if the setting actually changes.

Comment (joestringer) [reply Documentation/cmdref/cilium-agent.md]: Put another way, if users really are saving ~150MB of memory with this patch, then those users are also going to have connection disruption during upgrade.

Comment (borkmann) [reply Documentation/cmdref/cilium-agent.md]: Perhaps it may be better to have a tuning guide in our Cilium documentation where all these knobs could be described?

Comment (tklauser) [reply Documentation/cmdref/cilium-agent.md]: Sounds like it's better to abandon this PR and instead document theses knobs.

Comment (tgraf) [reply Documentation/cmdref/cilium-agent.md]: We need to decide on what is the ideal default and then apply that for new users while we point to the tuning guide for existing users. Right now, even tiny clusters are allocating a huge conntrack table.

Comment (tklauser) [reply Documentation/cmdref/cilium-agent.md]: As discussed in yesterday's Cilium community meeting, I'll add a note to the upgrade guide about this change and document how users can make sure they won't have disrupted connections during upgrade.

I'll also send a separate change which will allow to set the map sizes based on system memory size (probably in a separate PR) and document all the knobs users can use to optimize agent memory usage.

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-docs-please

**Comment by tklauser** [comment]
Added a section to the upgrade guide about the changed default value. I'm not entirely sure whether I got the part about dropped connections correct. @joestringer @tgraf @aanm could you please have a look?

**Comment by nebril** [comment]
test-me-please

File: Documentation/install/upgrade.rst
Comment (joestringer) [inline Documentation/install/upgrade.rst]:
```suggestion
   means at maximum 512K concurrent TCP connections across all endpoints are supported by default. To prevent from breaking established...
```

File: Documentation/install/upgrade.rst
Comment (joestringer) [inline Documentation/install/upgrade.rst]: This is tricky to get right. In general, we are providing instructions to users with the Helm options since v1.6, so the option here may be confusing because if someone applies it in Helm. For example, `--set ct-global...` it will have no effect.

I would suggest that instead we have a two-phase process:

1. Identify whether any action is necessary on the user's part. This could be either getting the logs from Cilium startup and checking the value (unreliable for long-running instances) or maybe using something like `bpftool` inside a cilium container to identify what the current size of the map is. We can likely streamline this to a 1-liner.
1. If (and only if) the value is non-standard, provide the helm options necessary to ensure that the `helm template` / `helm install` commands will generate the correct configuration for these settings.

**Review by joestringer — CHANGES_REQUESTED** [review]

File: Documentation/install/upgrade.rst
Comment (joestringer) [reply Documentation/install/upgrade.rst]: (note that in the code, TCP CT table is separate from non-TCP CT table so the limits are different).

File: Documentation/install/upgrade.rst
Comment (joestringer) [reply Documentation/install/upgrade.rst]: One other caveat here -- I didn't trace through the various install instructions but what subset of users will be affected here? I think we can safely say that:
* Users who installed v1.5, v1.6 or v1.7 on a fresh cluster and subsequently upgrade will not be affected unless they have explicitly configured these options in the past.
* Users who upgrade to newer Cilium versions while retaining their old ConfigMap (option B [here](https://docs.cilium.io/en/v1.7/install/upgrade/#step-2-option-b-preserve-configmap) *may* be affected (but it would depend on the initial version of Cilium used and the YAMLs provided with that version, and whether the user has made any manual changes to the ConfigMap since then..).

I would encourage you to investigate these statements and disagree if you find evidence that they're untrue.

The importance of this is that we can frame the documentation in such a way that we can potentially simplify the upgrade process for a plurality of users. Two focuses: (1) as a user, how much do you care about hitless upgrade / connection interruption (if you don't care, ignore these instructions) ; (2) how can a user know easily that they must take action (if you care, do you need to do something).

File: Documentation/install/upgrade.rst
Comment (tklauser) [reply Documentation/install/upgrade.rst]: Good point. ~I completely forgot to mention `bpf-nat-global-max` which is also reduced by this PR. Will adjust the docs as you suggest and also add a similar note about `bpf-nat-global-max`.~ (sorry, got confused about `bpf-nat-global-max`, it's derived from `bpf-ct-global-tcp-max`, so probably no need to mention it explicitly)

File: Documentation/install/upgrade.rst
Comment (tklauser) [reply Documentation/install/upgrade.rst]: Thanks a lot for the feedback @joestringer 

re. your 1st reply: I added the corresponding instructions to determine the current map size (using the `bpf-map` tool, as `bpftool` didn't seem to work for me on 4.9 kernel). Also added the instructions to set the helm options.

re. your 2nd reply: I traced through the install and upgrade instructions and as far as I can see, your statements given above are all true. But to be honest I'm not yet that familiar with helm and ConfigMap, so I might have missed some subtle detail.

> The importance of this is that we can frame the documentation in such a way that we can potentially simplify the upgrade process for a plurality of users. Two focuses: (1) as a user, how much do you care about hitless upgrade / connection interruption (if you don't care, ignore these instructions) ; (2) how can a user know easily that they must take action (if you care, do you need to do something).

Completely agree. I think having the instructions with the check for the currently configured map size should help in that regard. Maybe it can be fleshed out a bit more to be more concise. Feedback welcome.

**Comment by tklauser** [comment]
test-docs-please

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by tklauser** [comment]
test-me-please

File: Documentation/install/upgrade.rst
Comment (jrfastab) [reply Documentation/install/upgrade.rst]: @joestringer waiting for an ack on your side before I merge this <message from janitor hat>

File: Documentation/install/upgrade.rst
Comment (joestringer) [inline Documentation/install/upgrade.rst]: This sentence is misleading, in particular because it puts doubt on which value may cause connection breakage. To be clear, if any value for these settings is used that is not the same as the currently configured values, then there will be connection breakage.

I think it's also worth stating the impact: Connection breakage means that ongoing connections will be disrupted. Tolerant applications will reconnect after a blip. Less tolerant applications may require restarting to re-establish connectivity.

The paragraph that declares the impact should probably be separate from the other paragraphs; that way, if a user scans across the different paragraphs, they should be able to quickly identify that this is the paragraph stating the impact. They can make a judgement call whether they care in this case (and for that matter, get a hint what to do if things break during the upgrade).

File: Documentation/install/upgrade.rst
Comment (joestringer) [inline Documentation/install/upgrade.rst]: "or below" here is misleading in the same way as above; if the value is not the _same_ as the new configuration, then connectivity breakage will occur.

File: Documentation/install/upgrade.rst
Comment (joestringer) [inline Documentation/install/upgrade.rst]: FWIW if we wanted to avoid `bpf-map` we could just check the headers we use to compile the datapath:
```
# grep -R CT_MAP_SIZE_TCP /var/run/cilium/state/templates/
/var/run/cilium/state/templates/942ab6c4186d8b5f243c0f6fa59341122f671be1/lxc_config.h:#define CT_MAP_SIZE_TCP 524288
```

**Comment by joestringer** [comment]
test-docs-please

File: Documentation/cmdref/cilium-agent.md
Comment (joestringer) [inline Documentation/cmdref/cilium-agent.md:34]: I'm actually more worried about this setting affecting connectivity than the `--bpf-ct-global-tcp-max` setting. As far as I can tell, we never configured this using Helm so this is the default value for all deployments and all users who use the global NAT table may be affected by this. Seems like anyone who enables BPF nodeport / kube-proxy-free. Do we know the dataplane impact here during upgrade?

File: Documentation/install/upgrade.rst
Comment (joestringer) [reply Documentation/install/upgrade.rst]: I'm not sure how to tell the equivalent for the NAT table size since I didn't have an instance with BPF nodeport lying around to test. Probably `SNAT_MAPPING_IPV4_SIZE`?

**Review by joestringer — CHANGES_REQUESTED** [review]
Bear with me, I'm trying to ensure we're crystal clear here on the impact and making sure users come away with the right information when reading this :-) 

Overall, I think it's safe to assume that 95% of users are deploying Cilium via our Helm templates. From that perspective, I am mostly concerned about streamlining the instructions for those users. Obviously we care about users deploying in other ways as well, but there may be solutions that can move the complexity to a dedicated section if it doesn't affect most users. (I didn't make any specific recommendations below on this but that's my thinking while reviewing the PR).

Here are the two changes I see here and their potential impact:
* Changing `--bpf-global-tcp-max`. From my understanding, no user who deployed a fresh version of Cilium 1.5 or later via official instructions is affected. Users upgrading from older versions may be affected. Users who deploy via means other than the official Cilium pod may be affected.
* `--bpf-nat-global-max`. This one I don't understand as well. Seems to me like all BPF nodeport users may be affected by this, so I guess that's v1.6 or later. @brb  / @borkmann if this map is recreated, how would you characterize the impact? Same as CT, broken connections that may require apps to be restarted to resolve the issue?

This second one I suspect is not tested in the upgrade tests, which would hide the impact of this PR.

File: Documentation/install/upgrade.rst
Comment (joestringer) [reply Documentation/install/upgrade.rst]: Conversely... if the user previously configured the value to be something else, then they should hold on to their existing settings. Changing to 524288 or below will cause the same problem.

File: Documentation/install/upgrade.rst
Comment (tklauser) [reply Documentation/install/upgrade.rst]: Ok, my understanding was that only downsizing the map would cause connection breakage. I've now adjusted the section to state that any map size change will cause disrupted connections and also moved this statement into its own paragraph.

File: Documentation/install/upgrade.rst
Comment (tklauser) [reply Documentation/install/upgrade.rst]: > FWIW if we wanted to avoid `bpf-map` we could just check the headers we use to compile the datapath:
> 
> ```
> # grep -R CT_MAP_SIZE_TCP /var/run/cilium/state/templates/
> /var/run/cilium/state/templates/942ab6c4186d8b5f243c0f6fa59341122f671be1/lxc_config.h:#define CT_MAP_SIZE_TCP 524288
> ```

Good point, using that command instead.

Comment (tklauser) [reply Documentation/install/upgrade.rst]: > I'm not sure how to tell the equivalent for the NAT table size since I didn't have an instance with BPF nodeport lying around to test. Probably `SNAT_MAPPING_IPV4_SIZE`?

Yes, looks like it's `SNAT_MAPPING_IPV4_SIZE` and/or `SNAT_MAPPING_IPV6_SIZE` depending on IPv4/IPv6 being enabled/disabled. Not checked on an actual instance though, but by checking the sources.

Comment (tklauser) [reply Documentation/install/upgrade.rst]: Will add the corresponding `grep` command as well, if we decide to change the value of `bpf-nat-global-max`, see https://github.com/cilium/cilium/pull/10289#discussion_r389550453

File: Documentation/cmdref/cilium-agent.md
Comment (tklauser) [reply Documentation/cmdref/cilium-agent.md:34]: @borkmann @brb ^^^ do you know what the impact would be?

Given that the value is not configurable using Helm, maybe it's better to leave `bpf-nat-global-max` at the existing value for now and adjust it in a separate PR if needed?

Comment (borkmann) [reply Documentation/cmdref/cilium-agent.md:34]: We've sized the SNAT tables by default a bit smaller than the CT tables and similarly it should be done here (cc @brb ). Concern is that given SNAT cleanup depends on CT GC right now, then once CT LRU evicts entries, they go out of reach for SNAT GC. Hence 1mio vs 841429.

File: Documentation/install/upgrade.rst
Comment (borkmann) [reply Documentation/install/upgrade.rst]: `This means at maximum 512K concurrent TCP connections across all endpoints are supported by default.` This statement is actually not correct. There's a lot more added to the CT table, e.g. related entries, service entries, etc. So in the worse case in case of BPF NodePort when the backend is local, this means max 128k service connections. :-(

Comment (tklauser) [reply Documentation/install/upgrade.rst]: @borkmann TIL, thanks for clarifying. Do you think it's ok if I just remove this sentence? I think it would probably be too much detail in the upgrade guide to document the max. number of connections for all the cases.

File: Documentation/cmdref/cilium-agent.md
Comment (tklauser) [reply Documentation/cmdref/cilium-agent.md:34]: Thanks for the explanation. So I guess we should add a way to configure `bpf-nat-global-max` using Helm?

File: Documentation/install/upgrade.rst
Comment (borkmann) [reply Documentation/install/upgrade.rst]: Yes, I would remove it. Doesn't our helm deployment have a 512k CT table size for TCP anyway? Should there be any mentioned of this in the upgrade doc?

Comment (tklauser) [reply Documentation/install/upgrade.rst]: Yes, the helm deployment already has this number. I think the main concern is about users who changed the default value or aren't using helm deployment.

Comment (borkmann) [reply Documentation/install/upgrade.rst]: Right, should this be better clarified in the upgrade guide?

File: Documentation/cmdref/cilium-agent.md
Comment (borkmann) [reply Documentation/cmdref/cilium-agent.md:34]: Yes, I think it would be good. (Can be done in separate PR though)

File: Documentation/install/upgrade.rst
Comment (tklauser) [reply Documentation/install/upgrade.rst]: Agree, I'll rephrase the section to make it clearer.

File: Documentation/cmdref/cilium-agent.md
Comment (tklauser) [reply Documentation/cmdref/cilium-agent.md:34]: Will send a separate PR to allow configuring `bpf-nat-global-max` using Helm.

**Comment by tklauser** [comment]
Pushed a new version which should incorporate all the feedback regarding `bpf-ct-global-tcp-max`. As noted in https://github.com/cilium/cilium/pull/10289#discussion_r389617656 I'll send a separate PR for `bpf-nat-global-max` making it configurable using Helm and updating the upgrade guide to document it.

Please note that I changed the order of commits to make possible additional amendments easier to deal with.

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-docs-please

File: Documentation/cmdref/cilium-agent.md
Comment (tklauser) [reply Documentation/cmdref/cilium-agent.md:34]: Sent #10511 which will introduce the Helm option but leaving the value intact (i.e. at 841429) for now.

Since the value of `bpf-nat-global-max` is derived from `bpf-ct-global-tcp-max` and thus changed in this PR, I'll change the default value of `bpf-nat-global-max` for Helm in this PR as well.

File: Documentation/install/upgrade.rst
Comment (joestringer) [inline Documentation/install/upgrade.rst]: I think that the below should be accurate:
```suggestion
   effective in Cilium 1.6 or later, unless it was manually configured to a different
```

https://github.com/cilium/cilium/blob/215b0a49800e912d3cdeb70eb18e5bd42d6734fc/install/kubernetes/cilium/values.yaml#L191

File: Documentation/install/upgrade.rst
Comment (joestringer) [inline Documentation/install/upgrade.rst]: nit:
```suggestion
   ``bpf-ct-global-tcp-max`` in the daemon was set to the default value ``1000000`` to retain
```

File: Documentation/cmdref/cilium-agent.md
Comment (joestringer) [reply Documentation/cmdref/cilium-agent.md:34]: I didn't get a clear picture of what the datapath impact is here, but I'm presuming that we need similar upgrade docs handling for this option as for the CT TCP map option?

**Review by joestringer — CHANGES_REQUESTED** [review]
Almost there for CT table size upgrade notes, I think it's clearer now.

I still don't have a clear picture of the impact of `--bpf-nat-global-max` or the exact mitigating conditions (Only affects users with BPF nodeport enabled?)

**Comment by tklauser** [comment]
> Almost there for CT table size upgrade notes, I think it's clearer now.
> 
> I still don't have a clear picture of the impact of `--bpf-nat-global-max` or the exact mitigating conditions (Only affects users with BPF nodeport enabled?)

As discussed offline: I'll investigate this on a test setup and will add a separate section for `--bpf-nat-global-max` to the upgrade guide with my findings.

File: Documentation/install/upgrade.rst
Comment (joestringer) [inline Documentation/install/upgrade.rst]: By the way, I think that if the feature is not enabled, the above grep will output nothing; in that case the user should be safe to to nothing.

Comment (tklauser) [reply Documentation/install/upgrade.rst]: Agree. Already had this in my local changes but I seem to have pushed an intermediate version. Will still need to verify consequences for NAT table upgrade and will push once it's all documented.

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
Now verified in an actual setup with active NAT table (using the kube-proxy-free guide as per @brb's suggestion, thanks). I can confirm that similar countermeasures as for the CT map are necessary. I've updated the upgrade guide with the corresponding instructions.

PTAL, hopefully this can finally :rocket: :smile:

File: Documentation/cmdref/cilium-agent.md
Comment (tklauser) [reply Documentation/cmdref/cilium-agent.md:34]: Upgrade docs added in latest revision

**Comment by tklauser** [comment]
test-docs-please

**Review by joestringer — APPROVED** [review]
The new instructions are very clear and readable. :rocket:

**Pull request merged**
